### PR TITLE
[SVCS-580] Collapsible directories for zip rendering

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Ticket

https://openscience.atlassian.net/browse/SVCS-580

## Purpose
Our zip render currently lists files of all directories in one big table, other implementations have a neater collapsible hierarchical structure. This PR creates a collapsible directory UI.

## Changes

Uses two new JS libraries, jstree and jstree-table and method for the renderer that  reconstructs a file tree from a list of paths.

## Side effects

None that I know of

## QA Notes

Upload a complex zip file and click through it contents.

## Deployment Notes

None that I know of.

### CODE REVIEW

File: mfr/extensions/zip/templates/viewer.mako [hunk 1]
```diff
@@ -1,31 +1,43 @@
...
-                    <td>${file['name']}</td>
-                    <td>${file['date']}</td>
-                    <td>${file['size']}<td>
-                </tr>
-            % endfor
-        </tbody>
-    </table>
+<link rel="stylesheet" href="../static/css/bootstrap.min.css">
+<link rel="stylesheet" href="../static/css/default.css">
+<link rel="stylesheet" href="../static/jstree-theme/style.css"/>
```
Comment (AddisonSchiller) [inline mfr/extensions/zip/templates/viewer.mako]: You should be passing in `assets_url` and defining these as
`${base}/static/css......`
and not with `.../static`

File: mfr/extensions/zip/render.py [hunk 2]
```diff
@@ -17,15 +16,56 @@ class ZipRenderer(extension.BaseRenderer):
...
+                length = len(node_path['children'])
+                if not length or node_path['children'][length - 1]['text'] != path:
+
+                    new_node = dict(text=path)
+
+                    new_node['children'] = []
+
+                    if new_node['text']:
+                        date = '%d-%02d-%02d %02d:%02d:%02d' % file.date_time[:6]
+                        size = sizeof_fmt(int(file.file_size))
```
Comment (AddisonSchiller) [inline mfr/extensions/zip/render.py]: Folder size is always 0. Maybe folder sizes shouldn't be displayed. Has this been demo'd for product yet? See what they say.

File: mfr/extensions/zip/render.py [hunk 3]
```diff
@@ -17,15 +16,56 @@ class ZipRenderer(extension.BaseRenderer):
...
+                                new_node[
+                                    'icon'] = self.icons_url + '/generic-file.png'
+                        else:
+                            new_node['icon'] = self.icons_url + '/folder.png'
+
+                        node_path['children'].append(new_node)
+
+                    node_path = new_node
+                else:
+                    node_path = node_path['children'][length - 1]
```
Comment (AddisonSchiller) [inline mfr/extensions/zip/render.py]: Okay i understand the `if` statement a bit better, but there has to be a better way to write this.

File: mfr/extensions/zip/templates/viewer.mako [hunk 4]
```diff
@@ -1,31 +1,44 @@
...
-            % for file in zipped_filenames:
-                <tr>
-                    <td>${file['name']}</td>
-                    <td>${file['date']}</td>
-                    <td>${file['size']}<td>
-                </tr>
-            % endfor
-        </tbody>
-    </table>
+<link rel="stylesheet" href="${base}/css/bootstrap.min.css">
```
Comment (AddisonSchiller) [inline mfr/extensions/zip/templates/viewer.mako]: this shoud be down with `..bootstrap.min.js` and it shouldnt be defined by base, it should be defined like
`<link rel="stylesheet" href="/static/css/bootstrap.min.css">`

File: mfr/extensions/zip/render.py [hunk 5]
```diff
@@ -17,15 +16,56 @@ class ZipRenderer(extension.BaseRenderer):
...
-                    if not file.filename.startswith('__MACOSX')]
+        data = self.filelist_to_tree(files)
 
-        message = '' if filelist else 'This zip file is empty.'
+        return self.TEMPLATE.render(data=data, base=self.assets_url)
 
-        return self.TEMPLATE.render(zipped_filenames=filelist, message=message)
+    def filelist_to_tree(self, files):
+
+        self.icons_url = self.assets_url + '/img/icons'
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply mfr/extensions/zip/render.py]: Works fine for me.

**Comment by Johnetordoff** [comment]
@cslzchen  I've brought this up to date and consolidated the history.

**Comment by Johnetordoff** [comment]
Duplicated CSS Files:
- Was able to remove these.
Two Images:
- These are necessary to the render, why are they mapped into one file? I don't know, but it's not uncommon thing. 32px.png are the regular set, 40px.png for mobile.
One Script:
-  We do have a customization for this, I've added a comment.

**Comment by cslzchen** [comment]
Merged into a [feature branch](https://github.com/CenterForOpenScience/modular-file-renderer/tree/feature/improve-zip-rendering) where I will make some final tweaks. Consider this a merge and thanks for the great work 🎆.

**Pull request closed without merging**